apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs

```diff
@@ -73,29 +79,50 @@ public SimplePubSubTest(PubsubFixture fixture)
                     }
                 )).ConfigureAwait(false);
             var simpleSubscriber = await SimpleSubscriber.CreateAsync(subscriptionName,
-                settings: timeouts == null ? null : new SimpleSubscriber.Settings
+                settings: new SimpleSubscriber.Settings
                 {
-                    StreamAckDeadline = timeouts.Value
+                    StreamAckDeadline = timeouts,
+                    FlowControlSettings = new FlowControlSettings(maxMessagesInFlight, null)
                 }).ConfigureAwait(false);
 
             Console.WriteLine("Topic, Subscription, SimplePublisher and SimpleSubscriber all created");
 
             // Subscribe
             object recvLock = new object();
             int recvCount = 0; // Count of received messages
+            int dupCount = 0; // Count of duplicate messages
             long recvSum = 0L; // Sum of bytes of received messages
+            var recvedIds = new ConcurrentDictionary<int, bool>();
+            var nackedIds = new HashSet<int>();
             Task subTask = simpleSubscriber.StartAsync((msg, ct) =>
             {
-                var localRecvCount = Interlocked.Increment(ref recvCount);
-                if (localRecvCount <= initialNackCount)
+                int id = BitConverter.ToInt32(msg.Data.ToArray(), 0);
+                lock (nackedIds)
                 {
-                    return Task.FromResult(SimpleSubscriber.Reply.Nack);
+                    if (nackedIds.Count < initialNackCount)
+                    {
+                        if (nackedIds.Add(id))
+                        {
+                            // This ID not already nacked
+                            var localRecvCount = Interlocked.Increment(ref recvCount);
```

Comment:
What if this is cancelAfterRecvCount? Or is that always greater than initialNackCount? (Is that checked?)

Reply:
cancelAfterRecvCount only applies to non-nacked (acked) messages, although I agree that's not very clear.
(Also got rid of pointless allocation)